### ISSUE TITLE
Change buildChildList to work with proper arrays

### ISSUE THREAD

**Issue by drexkdal** [prescottprue/react-redux-firebase#324]
The function `buildChildList` that is invoked when populating an array (rather than a single property) expects an object instead of a proper array. I don't why the code is written like that, since the Firebase API automatically casts array-like data as an actual array.

Consequently, however, the lodash function `mapValues` is used, which returns an object (!) where the keys are array indices.

https://github.com/prescottprue/react-redux-firebase/blob/71a3e7781a77052afc7d0b3c979d385f66faad62/src/helpers.js#L248

I strongly suggest changing `mapValues` to a proper array `map` function, returning an array like you would expect for an array input.

**Comment by prescottprue** [comment]
@drexkdal It is also invoked when populating a key list (which is the reason for using `mapValues` since we intend to have an object come back).

Open to checking if it is an array or object before doing this, but the population of an object is *very common* so any solution would still need to support that (which switching to `map` would not).

**Comment by drexkdal** [comment]
I did notice the signature. Could you describe the use case for populating an object? What would the keys be?

It is surely a bad practice to populate an object with ids as both keys and values 🙂

From what you describe, it seems clear that using the same function for populating objects and arrays is incorrect – they should be delegated to each their own `buildChild*` function.

**Comment by prescottprue** [comment]
@drexkdal It is actually seen as a "Best Practice" on Firebase because it is common to store `$key: true` lists instead of arrays.

**Why?**
Removing an item from an array on Firebase can get annoying because using `update` with arrays is not supported. You have to get the whole array, remove the item locally, then call `set`. You can "guess" at the index, but that gets hairy when there are multiple clients that could "fight over" array indexes (using `transaction` can help).

All of that said, in my experience, as well as according to many Firebase docs, it is helpful to store things by key (often a push key).

**A Change**
Totally open to the pull request if you have the time. It could be published very quickly as the v2.0.0 pre-release cycle has been a release every few days (usually less than a week).

**Comment by drexkdal** [comment]
Thanks for the explanation. I still don't understand how `$key: true` is being populated, in practice, when it is the value, not the key, that is taken as an id (hence `mapValues`)?

**Comment by prescottprue** [comment]
@drexkdal Since we are mapping over an object of `$key: true`, we are hoping to map the value (which is currently `true`) to whatever we are populating with.

Consider the following structure of projects and users:

```js
{
  "projects": {
    "-KxsXreZGl4-ZZDQ3hc9": {
      "name": "Program JS",
      "createdAt": 1509555927968,
      "collaborators": {
        "00rnfZOHteSCY8y5oqJy1k1NTJu2": true,
        "018ixBKuIAVDgrqCMGl1MQUk2El2": true
      }
    }
  },
  "users": {
    "00rnfZOHteSCY8y5oqJy1k1NTJu2": {
      "displayName": "Rick Sanchez",
      "email": "rick@sanchez.com"
    },
    "018ixBKuIAVDgrqCMGl1MQUk2El2": {
      "displayName": "Morty Smith",
      "email": "morty@ohboy.com"
    },
  }
}
```

Populating using: `{ path: 'projects', populates: [
```
const populates = [{ child: 'collaborators', root: 'users' }]

firebaseConnect([
  { path: 'projects', populates }
]),
connect((state) => ({
  projects: populate(state.firebase, 'projects', populates)
})
```

Would yeild

```js
{
  "projects": {
    "-KxsXreZGl4-ZZDQ3hc9": {
      "name": "Program JS",
      "createdAt": 1509555927968,
      "collaborators": {
        "00rnfZOHteSCY8y5oqJy1k1NTJu2": {
          "displayName": "Rick Sanchez",
           "email": "rick@sanchez.com"
         },
         "018ixBKuIAVDgrqCMGl1MQUk2El2": {
           "displayName": "Morty Smith",
           "email": "morty@ohboy.com"
          },
      }
    }
  }
}
```

Notice how we are using `mapValues` to map the values of `collaborators`, which is an object, into an object by the same keys just with the values "populated".

Is there a use-case you have for wanting to change the signature of `buildChildList` or was this more just trying to understand what is going on? Populating arrays should be supported (let me know if you are having an issue), this is just an internal function that helps with the process.

For questions/discussion about internals, you may get a faster response if you reach out [over gitter](https://gitter.im/redux-firebase/Lobby).

Going to close. If there is a usecase for changing, we can reopen.

**Comment by drexkdal** [comment]
I understand the point about removing or updating items, but how is sequence preserved if it is important? Do people silently rely on the fact that JavaScript engines tend to retain declaration order, even though its not required by the language? Or do they store an array of ids separate from the `$key: true` map (which actually removes the point)?

Populating arrays doesn't work with the current code – they get converted into objects – so yes, there is a use case 🙂 I would reopen the issue.

**Comment by prescottprue** [comment]
@drexkdal There was a fix for this released in [`v2.0.0-beta.15`](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.0-beta.15), did you give that a try yet?

Arrays can also be used if preserving order is required. react-redux-firebase [uses the `state.firebase.ordered`](http://docs.react-redux-firebase.com/history/v2.0.0/docs/queries.html#ordered) to store ordered data (i.e an array instead of an object).

**Comment by drexkdal** [comment]
I've upgraded to `beta.15` now, but I still get an object back from `populate` instead of an array. I don't see any changes in that part of the code, so I guess it's not that strange.
